### ISSUE TITLE
Support RFC2253 X.500 output

### ISSUE THREAD

**Issue by NWilson** [Mbed-TLS/mbedtls#193]
The `x509_dn_gets` function is relative inflexible: it prints the X500 name in roughly the same format as OpenSSL's `X509_NAME_oneline` function.

In our application, we use RFC2253 for printing X.500 names. It's still not ideal because the format is ambiguous, but it gives us better interoperability between the various formats in the tools we use (OpenSSL clients, and JSSE server-side). Obviously, we try and avoid comparing string names wherever possible!

I've written a `x509_dn_gets_ext` function which prints out X.500 names, with a couple of flags so that you can get output that looks like JSSE, OpenSSL, or the current PolarSSL format. Please indicate if you'd be interested in a pull request for this.

> On a side note, I find the `polarssl_snprintf` usage very hard to follow/check. `SAFE_SNPRINTF` and `compat_snprintf` are copy-pasted into several files rather than pulled out properly into a header, which is nasty to begin with. Then, each bit of code has to work around the unreliable return codes, with messy conditionals. It shouldn't be so hard and brittle just to stick some strings together! I feel there should be a helper which handles that so the logic doesn't have to duplicated everywhere, avoiding the potential of a nasty off-by-one error in some little-used function.
> 
> I've created a simple string builder class for safely appending strings into a buffer, which is all that PolarSSL really needs. If you're interested, I'll finish the job and go through purging some more uses of `polarssl_snprintf` and replacing with the wrapper function.
> 
> On another side-note, I'm surprised that PolarSSL doesn't seem to have a function for transcoding an ASN.1 string into UTF-8! Certainly this would be a useful function for some people (ie anyone who wants to print it! A BMPString is _not_ ASCII.). OpenSSL's equivalent is `ASN1_STRING_to_UTF8`.

The code's currently sitting in a branch on my GitHub fork, in need of a bit of final cleaning up (needs test coverage!).

**Comment by mpg** [comment]
We might be interested in your extended function depending on its footprint (both in terms of code size, and foreseeable increase in maintenance effort). Could you please point me at the relevant branch in you fork so that I can evaluate that?

Regarding the current snprintf mess, we plan to normalize that in the next stable branch. A part of the fix will be to depend on PLATFORM_C on Windows to provide a standard-compliant wrapper for snprintf(), which will already makes things nicer, then we'll be able to make the rest more uniform. Again, I don't want to ask you to prepare a PR before I know if we're interested, so is there some place where I could look at your string build class?

Regarding BMPStrings, I'll have to check a reference, I'll get back to you later about that.

**Comment by NWilson** [comment]
Sure! I appreciate the footprint concern, one hundred 1% size increases would make it a bigger library. Happily, because I've replaced uses of inline code from `SAFE_SNPRINTF()` with a wrapper function, the library is actually 0.1% _smaller_ after my changes (stripped -O2 Release build).

My branch is here: https://github.com/NWilson/mbedtls/commits/rfc2253

I'm sorry the commit is a bit big, I got carried away, but could I split it up if that would help.

Unfortunately, the problem with the return value of `snprintf()` is worse than just Windows, it's actually a problem on HP-UX and older Solaris versions too (still within vendor support), and we support those platforms. What I've done is made a printf wrapper function which safely calls vsnprintf, but put an option in config.h POLARSSL_HAVE_STDARG to enable turning off support for vsnprintf, in which case inline code will be used to call snprintf directly. I admit it's not pretty, but it works and means that platforms without stdarg.h will continue to work (are there any?).

I know I would need to add some test cases to the test_suite_x509parse -- I'll spend the time on that if you say you're interested.

**Comment by mpg** [comment]
This commit is too big indeed :) Your size measurements are encouraging however!

I'm pressed with other work right now, but I'll get back to you probably next week about the `snprintf()` part, because I planned to work on it anyway. I might need a bit more time to get back to you about the other issues, though. Thanks in advance for you patience :)

If we're interested, I'll probably ask you to split the commit, but I wouldn't recommend you to do it now, because the current form is enough for reviewing and determining which parts we're interested in.

**Comment by mpg** [comment]
> Unfortunately, the problem with the return value of snprintf() is worse than just Windows, it's actually a problem on HP-UX and older Solaris versions too (still within vendor support), and we support those platforms.

I started doing some research on this, and it seems to me that the versions of Solaris with a non-compliant `snprintf()` are no longer supported: according to [this page](http://demin.ws/blog/english/2013/01/28/use-snprintf-on-different-platforms/) 5.10 is OK and according to Wikipedia, support for 5.9 ended in Oct 2014. So we have HP-UX and Windows, and since HP-UX seems to correctly 0-terminate, Windows is the only real problem.

**Comment by mpg** [comment]
If you're still interested in this, and when you find the time, I'm interested in having a proper pull request for that, based on our "development" branch, now that 2.0 is out. You'll find we rationalized our use of snprintf() and related work-arounds. Ideally your pull request would have separate commits for the following features (I didn't think about the order):

* proper string comparisons (minimal, we don't want to do Unicode normalization or so, but if we can accept more strings as equal when we should, while still rejecting all strings that we should reject, without much more code, then good).
* better string printing, converting everything to UTF-8 (again minimal I guess you get the idea)
* support for RFC 2253 format (maybe directly in x509_dn_gets(), unless you think it would break things - I'd like our default format to be a standard one)
* support for other formats in another if you think it's useful (should be a config.h option)

Generally speaking, we tend to avoid having our modules depend on other "common" modules (such as a string builder). On the other hand, if the string builder is only used in X.509, it is probably not a problem as X.509 already has so many dependencies, so we might go for it, if it makes the code simpler and/or the binary more compact. @pjbakkker feel free to chime in on that point :)

**Comment by NWilson** [comment]
Hi, thanks for your comments, and sorry for the very slow reply.

I do plan to come back to this, at some point, with a better PR. This has dropped down my list of priorities, and I might be able to find some time to return to it in October.

**Comment by mpg** [comment]
No problem, we've been pretty busy too :) A PR is still welcome obviously, if/when you have the time to prepare it.

**Comment by davidhorstmann-arm** [comment]
Closing as superseded by #6785.
